onnxruntime/core/framework/graph_partitioner.cc

```diff
@@ -533,6 +557,78 @@ static Status InlineNodes(Graph& graph, bool& modified_graph) {
   return Status::OK();
 }
 
+static Status IntelligentAotFunctionInliningImpl(const ExecutionProviders& execution_providers,
+                                                 const KernelRegistryManager& kernel_registry_mgr,
+                                                 Graph& graph,
+                                                 size_t& inlined_round_count) {
+  // handle testing edge case where optimizers or constant lifting results in graph with no nodes.
+  // doing it here saves all providers checking for this in GetCapability
+  if (graph.NumberOfNodes() == 0) {
+    return Status::OK();
+  }
+
+  for (auto& node : graph.Nodes()) {
+    for (auto& entry : node.GetAttributeNameToMutableSubgraphMap()) {
+      Graph* subgraph = entry.second;
+      // we pass through the FuncManager from the top level graph
+      ORT_RETURN_IF_ERROR(IntelligentAotFunctionInliningImpl(execution_providers,
+                                                             kernel_registry_mgr,
+                                                             *subgraph,
+                                                             inlined_round_count));
+    }
+  }
+
+  // Gather the candidates
+  InlinedVector<NodeIndex> inline_candidates;
+  for (auto& node : graph.Nodes()) {
+    if (node.CanBeInlined()) {
+      inline_candidates.push_back(node.Index());
+    }
+  }
+
+  if (inline_candidates.empty()) {
+    return Status::OK();
+  }
+
+  // Find out all the nodes that are already taken
+  InlinedHashSet<NodeIndex> claimed_by_ep;
+  for (const auto& ep : execution_providers) {
+    std::vector<std::unique_ptr<ComputeCapability>> capabilities;
+    ORT_RETURN_IF_ERROR(GetCapabilityForEPForAotInlining(graph, kernel_registry_mgr, *ep, capabilities));
```

Comment:
I am wondering a bit about how this logic has been split up (between this loop and GetCapabilityForEPForAotLining). Is there any plan to allow EPs to customize/override this later on? Reusing the existing GetCapability makes sense (for reducing the impact on EPs). But was wondering.

Reply:
Just to continue on the above point: it seems to me that the ideal API would be a function that returns a set of Nodes to be inlined. We could have a default implementation for that based on reusing the existing GetCapability api (and this logic seems to be currently split up here).

But I do have one question about the loop below (which comes from a similar loop in the partitioning algorithm, I assume): the idea seems to be that when an EP X asks for a subgraph, and even a node in that subgraph is already assigned to some other EP, EP X does not get any of the nodes in that subgraph. I don't quite understand this mechanism, especially in this context. It may be likely that we will never run into this situation in practice (since it will require 2 EPs other than the CPU EP) ... not sure if that is common, though I guess using both CUDA EP and TensorRT EP is possible.

Reply:
We assume that the partitioning algorithm would make EP assignments based on the same logic, and currently logic is as implemented. If wanted to change that, we probably should change in both places.

Reply:
The point of concern is noted. It calls for a broader discussion.